subsys/bluetooth/mesh/vnd/dm_srv.c

```diff
@@ -0,0 +1,474 @@
+
+rsp:
+	result_status_send(model, ctx, 1, err);
+	return -err;
+}
+
+static int handle_result_get(struct bt_mesh_model *model, struct bt_mesh_msg_ctx *ctx,
+			     struct net_buf_simple *buf)
+{
+	struct bt_mesh_dm_srv *srv = model->user_data;
+	uint8_t cnt = net_buf_simple_pull_u8(buf);
+
+	if (!cnt) {
+		result_status_send(model, ctx, cnt, EBADMSG);
+		return 0;
+	}
+
+	if (!srv->results.available_entries) {
+		result_status_send(model, ctx, cnt, ENOENT);
+		return 0;
+	}
+
+	result_status_send(model, ctx, cnt, 0);
+
+	return 0;
+}
+
+static int handle_sync(struct bt_mesh_model *model, struct bt_mesh_msg_ctx *ctx,
+		       struct net_buf_simple *buf)
+{
+	int err;
+	struct bt_mesh_dm_srv *srv = model->user_data;
+
+	if (srv->is_busy) {
+		return -EBUSY;
+	}
+
+	srv->target_addr = ctx->addr;
+
+	enum dm_ranging_mode method = net_buf_simple_pull_u8(buf);
+	uint8_t timeout = net_buf_simple_pull_u8(buf);
+	struct dm_request req = {
+		.role = DM_ROLE_INITIATOR,
+		.ranging_mode = method,
+		.rng_seed = (ctx->addr << 16) + ctx->recv_dst,
+		.start_delay_us = 0
+	};
+
+	srv->is_busy = true;
+	k_sem_reset(&srv->dm_ready_sem);
+	err = dm_request_add(&req);
+	if (err) {
+		BT_ERR("DM add request failed: %d", err);
+		srv->is_busy = false;
+		return err;
+	}
+
+	srv->current_role = DM_ROLE_INITIATOR;
+	k_work_reschedule(&srv->timeout, K_MSEC(timeout * 100));
+
+	return 0;
+}
+
+static int handle_msg(struct bt_mesh_model *model, struct bt_mesh_msg_ctx *ctx,
+		      struct net_buf_simple *buf)
+{
+	uint8_t opcode = net_buf_simple_pull_u8(buf);
+
+	switch (opcode) {
+	case BT_MESH_DM_CONFIG_OP:
+		return handle_cfg(model, ctx, buf);
+	case BT_MESH_DM_START_OP:
+		return handle_start(model, ctx, buf);
+	case BT_MESH_DM_RESULT_GET_OP:
+		return handle_result_get(model, ctx, buf);
+	case BT_MESH_DM_SYNC_OP:
+		return handle_sync(model, ctx, buf);
+	}
+
+	return -EOPNOTSUPP;
+}
+
+const struct bt_mesh_model_op _bt_mesh_dm_srv_op[] = {
+	{
+		BT_MESH_DM_SRV_OP,
+		BT_MESH_LEN_MIN(BT_MESH_DM_CONFIG_MSG_LEN_MIN),
+		handle_msg,
+	},
+	BT_MESH_MODEL_OP_END,
+};
+
+static void data_ready(struct dm_result *result)
+{
+	if (!dm_srv->is_busy) {
+		BT_WARN("Unexpected DM result ready, discarding");
+		return;
+	}
+
+	BT_INFO("New DM result ready for target node %04x", dm_srv->target_addr);
+
+	new_entry_store(result);
+	if (k_work_delayable_is_pending(&dm_srv->timeout)) {
+		k_work_cancel_delayable(&dm_srv->timeout);
+		if (dm_srv->current_role == DM_ROLE_REFLECTOR) {
+			result_status_send(dm_srv->model, &dm_srv->rsp_ctx, 1, 0);
+		}
+	}
+
+	dm_srv->is_busy = false;
+}
+
+static struct dm_cb dm_cb = {
+	.data_ready = data_ready,
+};
+
+static void timeout_work(struct k_work *work)
+{
+	BT_ERR("DM attempt timed out");
+	if (dm_srv->current_role == DM_ROLE_REFLECTOR) {
+		result_status_send(dm_srv->model, &dm_srv->rsp_ctx, 1, ETIMEDOUT);
+	}
+
+	struct dm_result result = {
+		.quality = DM_QUALITY_DO_NOT_USE,
+		.status = false
+	};
+
+	new_entry_store(&result);
```

Comment:
We should probably store the result first and _then_ send the response? Also, this is more of a spec comment, but filling up the results array with failed measurements like this does not seem very optimal to me.